crates/pico_macros/src/db_macro.rs

```diff
@@ -1,52 +1,123 @@
+use convert_case::{Case, Casing};
 use proc_macro::TokenStream;
-use quote::quote;
-use syn::{DeriveInput, parse_macro_input};
+use quote::{format_ident, quote};
+use syn::{ItemStruct, Type, parse_macro_input};
 
 pub(crate) fn db_macro(item: TokenStream) -> TokenStream {
-    let input = parse_macro_input!(item as DeriveInput);
-    let struct_name = input.ident.clone();
+    let input = parse_macro_input!(item as ItemStruct);
+    let struct_ident = input.ident.clone();
     let (impl_generics, type_generics, where_clause) = input.generics.split_for_impl();
 
+    let fields = match &input.fields {
+        syn::Fields::Named(named) => &named.named,
+        _ => {
+            return syn::Error::new_spanned(&input.fields, "#[db] requires named fields")
+                .to_compile_error()
+                .into();
+        }
+    };
+
+    let mut getter_impls = Vec::new();
+    let mut aux_struct_defs = Vec::new();
+
+    for f in fields.iter() {
+        let field_ident = f.ident.as_ref().unwrap();
+
+        if field_ident == "storage" {
+            continue;
+        }
+
+        if is_phantom_data(&f.ty) {
+            continue;
+        }
+
+        let field_ty = &f.ty;
+        let counter_ident =
+            format_ident!("__{}Counter", field_ident.to_string().to_case(Case::Pascal));
+        let get_ident = format_ident!("get_{field_ident}_tracked");
+        let get_ident_mut = format_ident!("get_{field_ident}_tracked_mut");
+
+        aux_struct_defs.push(quote! {
+            #[derive(Clone, Copy, Default, PartialEq, Eq, Hash, ::pico_macros::Singleton)]
+            struct #counter_ident(u64);
+            impl #counter_ident {
+                #[inline]
+                fn increment(self) -> Self { Self(self.0.wrapping_add(1)) }
+            }
+        });
+
+        getter_impls.push(quote! {
+            #[inline]
+            pub fn #get_ident(&self) -> &#field_ty {
+                let _ = self.storage.get_singleton::<#counter_ident>();
+                &self.#field_ident
+            }
+
+            #[inline]
+            pub fn #get_ident_mut(&mut self) -> &mut #field_ty {
+                let next = self.storage.get_singleton::<#counter_ident>().copied().unwrap_or_default().increment();
```

Comment:
`.map(|x| x.increment())` is oh-so-slightly more efficient! No need to increment if its the first access